File: libs/core/src/private/syntax/token_stream.cpp
/**
 * @file token_stream.cpp
 * @author Francesco Corso
 * @date 7/12/2026
 * @brief
 */
module;

#include <libassert/assert-macros.hpp>

module prism.core:syntax.token_stream.impl;

import :syntax.token_stream;

namespace prism
{

    bool TokenStream::at_end()
    {
        return peek().kind() == SyntaxKind::eof_token;
    }

    const GreenToken &TokenStream::peek(const std::size_t count)
    {
        if (lookahead_.empty())
            buffer_tokens();

        return !lookahead_.empty() ? *lookahead_[count - 1] : previous();
    }

    void TokenStream::advance()
    {
        auto &token = peek();
        tokens_.push_back(token.shared_from_this());
        lookahead_.pop_front();
    }

    void TokenStream::replace_next(GreenPtr<GreenToken> token)
    {
        [[maybe_unused]] const auto &next = peek();
        DEBUG_ASSERT(next.kind() != SyntaxKind::eof_token);

        lookahead_.pop_front();
        lookahead_.push_front(std::move(token));
    }

    void TokenStream::buffer_tokens()
    {
        static constexpr std::size_t max_tokens = 1024;

        for ([[maybe_unused]] auto i : std::views::iota(0uz, max_tokens))
        {
            if (const auto &token = lookahead_.emplace_back(lexer_.next()); token->kind() == SyntaxKind::eof_token)
                break;
        }
    }
} // namespace prism

File: src/Prism.Core/Parser/TokenStream.cs
using System.Diagnostics;
using Nito.Collections;
using Prism.Core.Syntax;
using Prism.Core.Syntax.Green;

namespace Prism.Core.Parser;

internal sealed class TokenStream(string text)
{
    private readonly Lexer _lexer = new(text);
    private readonly List<GreenToken> _tokens = [];
    private readonly Deque<GreenToken> _lookahead = [];

    public bool AtEnd => Peek().Kind == SyntaxKind.EofToken;

    public IReadOnlyList<GreenToken> Tokens => _tokens;

    public GreenToken Previous
    {
        get
        {
            Debug.Assert(_tokens.Count > 0);
            return _tokens[^1];
        }
    }

    public GreenToken Peek(int count = 1)
    {
        if (_lookahead.Count == 0)
        {
            BufferTokens();
        }

        return _lookahead.Count > 0 ? _lookahead[count - 1] : Previous;
    }

    public GreenToken Consume()
    {
        var token = Peek();
        _tokens.Add(token);
        _lookahead.RemoveFromFront();
        return token;
    }

    public void Advance()
    {
        var token = Peek();
        _tokens.Add(token);
        _lookahead.RemoveFromFront();
    }

    public void ReplaceNext(GreenToken token)
    {
        var next = Peek();
        Debug.Assert(next.Kind != SyntaxKind.EofToken);

        _lookahead.RemoveFromFront();
        _lookahead.AddToFront(token);
    }

    public void ReplaceNext(params ReadOnlySpan<GreenToken> tokens)
    {
        var next = Peek();
        Debug.Assert(next.Kind != SyntaxKind.EofToken);

        _lookahead.RemoveFromFront();
        foreach (var token in tokens)
        {
            _lookahead.AddToFront(token);
        }
    }

    private void BufferTokens()
    {
        const int maxTokens = 1024;
        for (var i = 0; i < maxTokens; i++)
        {
            var token = _lexer.Next();
            _lookahead.AddToBack(token);
            if (token.Kind == SyntaxKind.EofToken)
                break;
        }
    }
}
